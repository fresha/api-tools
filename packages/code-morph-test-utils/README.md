# @fresha/code-morph-test-utils

This package contains testing utilities used by "code morphing" packages.

## Installation

```bash
$ npm install @fresha/code-morph-test-utils
```
