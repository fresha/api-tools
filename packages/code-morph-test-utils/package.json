{
  "name": "@fresha/code-morph-test-utils",
  "version": "0.1.1",
  "description": "Testing utilites for code morphers",
  "main": "build/index.js",
  "types": "build/index.d.ts",
  "scripts": {
    "build": "tsc --project tsconfig.build.json",
    "build:watch": "npm run build -- --watch",
    "check": "run-s lint test",
    "check:fix": "run-s lint:fix test",
    "clean": "rimraf ./build",
    "eslint": "eslint ./src",
    "eslint:fix": "eslint ./src --fix",
    "lint": "run-s eslint typecheck",
    "lint:fix": "run-s eslint:fix typecheck",
    "prebuild": "npm run clean",
    "test": "jest --passWithNoTests",
    "typecheck": "tsc --noEmit"
  },
  "keywords": [
    "code morphing",
    "testing",
    "Jest"
  ],
  "files": [
    "build/",
    "package.json",
    "LICENSE",
    "README.md"
  ],
  "publishConfig": {
    "access": "public"
  },
  "author": "Fresha Engineering",
  "contributors": [
    {
      "name": "Andriy Mykulyak",
      "email": "andriy@fresha.com",
      "url": "https://github.com/mykulyak"
    }
  ],
  "maintainers": [
    {
      "name": "Andriy Mykulyak",
      "email": "andriy@fresha.com",
      "url": "https://github.com/mykulyak"
    }
  ],
  "license": "MIT",
  "devDependencies": {
    "@fresha/eslint-config": "^0.1.1",
    "@fresha/typescript-config": "^0.1.1",
    "@types/jest": "^29.1.2",
    "typescript": "^4.7.2"
  },
  "dependencies": {
    "@fresha/jest-config": "^0.1.1",
    "jest-matcher-utils": "^29.1.2",
    "prettier": "^2.7.1"
  }
}
