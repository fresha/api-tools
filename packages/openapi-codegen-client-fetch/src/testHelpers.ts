// eslint-disable-next-line import/no-extraneous-dependencies
import { createTSProjectTestContext } from '@fresha/openapi-codegen-test-utils';

import type { ActionContext, Context } from './context';
import type { OpenAPIModel, OperationModel } from '@fresha/openapi-model/build/3.0.3';

export const createTestContext = (openapi: OpenAPIModel): Context => {
  const base = createTSProjectTestContext(openapi);

  return {
    ...base,
    includeDeprecated: false,
    includedTags: new Set<string>(),
    excludedTags: new Set<string>(),
    transformRequest: 'kebab',
    transformResponse: 'camel',
  };
};

export const createActionTestContext = (
  operation: OperationModel,
  fileName: string,
): ActionContext => {
  const base = createTestContext(operation.root);
  const sourceFile = base.project.createSourceFile(fileName, '');

  return {
    ...base,
    sourceFile,
    operation,
  };
};
