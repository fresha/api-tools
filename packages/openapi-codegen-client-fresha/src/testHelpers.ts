// eslint-disable-next-line import/no-extraneous-dependencies
import { createTSProjectTestContext } from '@fresha/openapi-codegen-test-utils';
import { getAPIName, getRootUrlOrThrow } from '@fresha/openapi-codegen-utils';

import type { Context } from './context';
import type { OpenAPIModel } from '@fresha/openapi-model/build/3.0.3';

export const createTestContext = (openapi: OpenAPIModel, rootDir = '/'): Context => {
  const base = createTSProjectTestContext(openapi, rootDir);

  let apiRootUrl: string;
  try {
    apiRootUrl = getRootUrlOrThrow(openapi);
  } catch (err) {
    apiRootUrl = 'API_URL';
  }

  return {
    ...base,
    apiName: getAPIName(openapi),
    apiRootUrl,
  };
};
