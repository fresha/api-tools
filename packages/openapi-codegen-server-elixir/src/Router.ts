import console from 'console';

import type { Controller } from './Controller';
import type { Generator } from './Generator';

export class Router {
  static makeFileName(): string {
    return '';
  }

  private readonly generator: Generator;

  constructor(generator: Generator) {
    this.generator = generator;
  }

  // eslint-disable-next-line class-methods-use-this
  processController(controller: Controller): void {
    console.log(!!controller);
  }

  generateCode(): void {
    console.log(this.generator);
  }
}
