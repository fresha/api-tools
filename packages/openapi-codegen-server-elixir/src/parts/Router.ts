import { maybeAddLeadingSlash } from '@fresha/openapi-codegen-utils';

import type { Controller } from './Controller';
import type { Context } from './types';

type Entry = {
  urlExp: string;
  controllerModuleAlias: string;
  actionNames: string[];
};

export class Router {
  readonly context: Context;
  protected readonly entries: Entry[];

  constructor(context: Context) {
    this.context = context;
    this.entries = [];
  }

  collectData(controller: Controller): void {
    this.context.logger.info(`Collecting routes for "${controller.moduleName}" controller`);

    const actionNames: string[] = [];
    for (const [, action] of controller.actionEntries()) {
      actionNames.push(action.getName());
    }

    this.entries.push({
      urlExp: controller.urlExp,
      controllerModuleAlias: controller.moduleAlias,
      actionNames,
    });
  }

  generateCode(): void {
    this.context.logger.info('Generating routing code');

    const lines = [
      '',
      '#',
      '# ROUTER',
      '# In your router file, add the following lines',
      '#',
      `scope "/api", ${this.context.project.getAppModuleName()} do`,
      '  pipe_through :api',
      '',
    ];

    for (const entry of this.entries) {
      const actions = entry.actionNames.map(a => `:${a}`).join(', ');
      lines.push(
        `  resources("${maybeAddLeadingSlash(entry.urlExp)}", ${
          entry.controllerModuleAlias
        }, only: [${actions}])`,
      );
    }

    lines.push('end', '');

    this.context.consoleWriter(lines.join('\n'));
  }
}
