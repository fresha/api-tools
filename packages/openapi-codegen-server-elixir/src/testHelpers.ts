import { Project } from '@fresha/code-morph-ex';
import { createRegistry } from '@fresha/json-api-model';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createTestContext as createBaseTestContext } from '@fresha/openapi-codegen-test-utils';
import { OpenAPIFactory } from '@fresha/openapi-model/build/3.0.3';

import { Generator } from './parts';

import type { Context } from './context';

export const createTestContext = (phoenixApp: string, rootDir = '/'): Context => {
  const openapi = OpenAPIFactory.create();
  const base = createBaseTestContext(openapi, rootDir);

  const project = new Project({
    rootDir,
    phoenixApp,
    useInMemoryFileSystem: true,
  });

  return {
    ...base,
    testObjectFactoryModuleName: `${project.getAppModuleName()}.Factory`,
    openapi,
    project,
    registry: createRegistry(),
  };
};

export const createGenerator = (phoenixApp = 'api_tools_web'): Generator => {
  const context = createTestContext(phoenixApp, '/');
  const generator = new Generator(context);
  return generator;
};
