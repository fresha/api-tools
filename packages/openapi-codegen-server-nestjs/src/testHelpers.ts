import { Nullable } from '@fresha/api-tools-core';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createTSProjectTestContext } from '@fresha/openapi-codegen-test-utils';
import { OpenAPIFactory, SchemaModel } from '@fresha/openapi-model/build/3.0.3';

import { Context } from './context';
import { Generator } from './Generator';

export const createTestContext = (nestApp = 'web', outputPath = '/tmp/'): Context => {
  const openapi = OpenAPIFactory.create();
  const base = createTSProjectTestContext(openapi, outputPath);

  base.project.createSourceFile(
    `${outputPath}/${nestApp}.module.ts`,
    `import { Module } from '@nestjs/common';
    import { AppController } from './app.controller';
    import { AppService } from './app.service';

    @Module({
      imports: [],
      controllers: [AppController],
      providers: [AppService],
    })
    export class AppModule {}`,
  );

  return {
    ...base,
    nestApp,
    addDTO: jest.fn(),
  };
};

export const createGenerator = (nestApp = 'app', rootDir = '/'): Generator => {
  const context = createTestContext(nestApp, rootDir);

  let generator: Generator;
  // eslint-disable-next-line prefer-const
  generator = new Generator({
    ...context,
    addDTO: (name: string, schema: Nullable<SchemaModel>) => generator.addDTO(name, schema),
  });

  return generator;
};
