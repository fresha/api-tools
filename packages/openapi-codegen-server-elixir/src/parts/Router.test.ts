import { poorMansElixirFormat } from '@fresha/code-morph-test-utils';

import { makeContext } from '../testHelpers';

import { Router } from './Router';

import type { Action } from './Action';
import type { Controller } from './Controller';

test('no paths', () => {
  const context = makeContext('api_tools_router_web');

  const router = new Router(context);
  router.generateCode();

  const { calls } = (context.console.log as jest.Mock).mock;
  expect(calls).toHaveLength(1);
  expect(poorMansElixirFormat((calls as string[][])[0][0])).toBe(
    poorMansElixirFormat(`
    #
    # ROUTER
    # In your router file, add the following lines
    #
    scope "/api", ApiToolsRouterWeb do
      pipe_through :api

    end
  `),
  );
});

test('multiple controllers and actions', () => {
  const context = makeContext('api_tools_router_web');

  const router = new Router(context);

  router.collectData({
    urlExp: '/users',
    moduleName: 'ApiToolsRouterWeb.UsersController',
    moduleAlias: 'UsersController',
    actionEntries(): IterableIterator<[string, Action]> {
      const actions = new Map<string, Action>([
        [
          'index',
          {
            getName() {
              return 'index';
            },
          } as Action,
        ],
        [
          'show',
          {
            getName() {
              return 'show';
            },
          } as Action,
        ],
      ]);
      return actions.entries();
    },
  } as Controller);

  router.collectData({
    urlExp: '/users',
    moduleName: 'ApiToolsRouterWeb.TasksController',
    moduleAlias: 'TasksController',
    actionEntries(): IterableIterator<[string, Action]> {
      const actions = new Map<string, Action>([
        [
          'delete',
          {
            getName() {
              return 'delete';
            },
          } as Action,
        ],
        [
          'create',
          {
            getName() {
              return 'create';
            },
          } as Action,
        ],
      ]);
      return actions.entries();
    },
  } as Controller);

  router.generateCode();

  const { calls } = (context.console.log as jest.Mock).mock;
  expect(calls).toHaveLength(1);
  expect(poorMansElixirFormat((calls as string[][])[0][0])).toBe(
    poorMansElixirFormat(`
    #
    # ROUTER
    # In your router file, add the following lines
    #
    scope "/api", ApiToolsRouterWeb do
      pipe_through :api

      resources("/users", UsersController, only: [:index, :show])
      resources("/users", TasksController, only: [:delete, :create])
    end
  `),
  );
});
