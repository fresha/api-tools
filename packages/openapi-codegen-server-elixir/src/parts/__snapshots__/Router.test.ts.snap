// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multiple controllers and actions 1`] = `
"
#
# ROUTER
# In your router file, add the following lines
#
scope "/api", ApiToolsRouterWeb do
  pipe_through :api

  resources("/users", UsersController, only: [:index, :show])
  resources("/users", TasksController, only: [:delete, :create])
end
"
`;

exports[`no paths 1`] = `
"
#
# ROUTER
# In your router file, add the following lines
#
scope "/api", ApiToolsRouterWeb do
  pipe_through :api

end
"
`;
